_posts/2015-09-01-Cocoa-Architecture-Dropped-Design-Patterns.markdown

```diff
@@ -0,0 +1,148 @@
+
+One pattern for handling this is [class composition](http://stackoverflow.com/questions/9710411/ios-grasping-composition). This is the idea that instead of having one object do multiple things, you allow a collection of objects to work together. Providing more space for each object to conform to the Single Responsibility Principal ([SRP](https://en.wikipedia.org/wiki/Single_responsibility_principle).) If you're interested in this, you may also be interested in the [class clusters](https://developer.apple.com/library/ios/documentation/General/Conceptual/CocoaEncyclopedia/ClassClusters/ClassClusters.html) pattern.
+
+A good example of this comes [from Energy](https://github.com/artsy/energy/blob/aa97d90cf37932d4c0f49ea4c4d31f7e491f16a6/Classes/Controllers/Top%20View%20Controller/ARTopViewToolbarController.m), our root view controller `ARTopViewController` used to control its own toolbar items. Over 4 years this became difficult to manage, and a lot of extra code in the view controller. By abstracting out the implementation details of managing the toolbar items into it's [own class](https://github.com/artsy/energy/blob/aa97d90cf37932d4c0f49ea4c4d31f7e491f16a6/Classes/Controllers/Top%20View%20Controller/ARTopViewToolbarController.m) we were able to allow the `ARTopViewController` to state what it wanted by not how it was done.
+
+## Configuration Classes over Inter-Class Communication
+
+A one of the most important aspects of Energy is to [email artworks](http://folio.artsy.net). So there is a lot of code around configuring the email you want to send, and then generating HTML from those settings. This started out pretty simple as we had very few app-wide settings. Over time, deciding what we need to show in terms of settings and how they affected the email became very complicated.
+
+The part that eventually became a strong enough code-smell to warrant a re-write was the view controller which allowed a partner to choose what information to send would pass details individually to an object that would generate the HTML. I found it difficult to write simple tests for the class' behavior. Initially I would mock the email composer, then inspect the methods that were called. This felt wrong, as you shouldn't really be mocking classes you own. Given the importance of the functionality that classes provide our application, ideas on ways to improve the section of code stayed on my mind for a long time.
```

Comment:
> The part that eventually became a strong enough code-smell to warrant a re-write was the view controller which allowed a partner to choose what information to send**. It** would pass details individually to an object that would generate the HTML.

Is that correct? I don’t quite understand what it’s explaining, though, especially that last sentence.